File: kotlin/src/main/kotlin/22.GenerateParentheses.kt
import org.junit.Assert.assertEquals
import org.junit.Test
// https://leetcode.com/problems/generate-parentheses/

class GenerateParenthesis {
  fun generateParenthesis(n: Int): List<String> {
    val result = mutableListOf<String>()

    fun build(pre: String, nl: Int){
      if(pre.length == n+n)
        result.add(pre)
      else{
        if(nl < n)
          build(pre+'(', nl+1)
        if(nl > pre.length - nl)
          build(pre+')', nl)
      }
    }

    if(n > 0) build("", 0)
    return result
  }


  @Test
  fun test1(){
    assertEquals(listOf(
      "((()))",
      "(()())",
      "(())()",
      "()(())",
      "()()()"
    ), generateParenthesis(3))
  }
}

File: python/22_GenerateParentheses.py
from unittest import TestCase
# https://leetcode.com/problems/generate-parentheses/


class GenerateParenthesis(TestCase):
    def generateParenthesis(self, n):
        result = []

        def build(pre, nl):
            if len(pre) == n+n:
                result.append(pre)
            else:
                if nl < n:
                    build(pre+"(", nl+1)
                if nl > len(pre) - nl:
                    build(pre+")", nl)
        if n > 0:
            build("", 0)
        return result

    def test1(self):
        self.assertEqual([
            "((()))",
            "(()())",
            "(())()",
            "()(())",
            "()()()"
        ], self.generateParenthesis(3))


